apps/hash-frontend/src/pages/settings/integrations/google-sheets/create-or-edit-sheets-integration.tsx

```diff
@@ -0,0 +1,412 @@
+        "https://hash.ai/@google/types/property-type/account-id/"
+      ],
+    query: currentIntegration?.query.properties[
+      "https://blockprotocol.org/@hash/types/property-type/query/"
+    ] as MultiFilter | undefined,
+  });
+
+  const [createEntity] = useMutation<
+    CreateEntityMutation,
+    CreateEntityMutationVariables
+  >(createEntityMutation);
+
+  const [updateEntity] = useMutation<
+    UpdateEntityMutation,
+    UpdateEntityMutationVariables
+  >(updateEntityMutation);
+
+  const { propertyTypes } = usePropertyTypes({ latestOnly: true });
+  const { latestEntityTypes } = useLatestEntityTypesOptional();
+
+  const entityTypeSchemas = useMemo(
+    () => latestEntityTypes?.map((type) => type.schema) ?? [],
+    [latestEntityTypes],
+  );
+
+  const propertyTypeSchemas = useMemo(
+    () => Object.values(propertyTypes ?? {}).map((type) => type.schema),
+    [propertyTypes],
+  );
+
+  const authContext = useGoogleAuth();
+  if (authContext.loading) {
+    return null;
+  }
+
+  const { addGoogleAccount, checkAccessToken, getAccessToken } = authContext;
+
+  const submittable =
+    !!integrationData.googleAccountId &&
+    !!integrationData.query &&
+    (integrationData.existingFile || integrationData.newFileName);
+
+  const submit = async () => {
+    if (
+      !integrationData.googleAccountId ||
+      !integrationData.query ||
+      (!integrationData.existingFile && !integrationData.newFileName)
+    ) {
+      return;
+    }
+
+    try {
+      await checkAccessToken({
+        googleAccountId: integrationData.googleAccountId,
+      });
+    } catch {
+      setShowReauthModal(true);
+      return;
+    }
+
+    let queryEntityId = currentIntegration?.query.metadata.recordId.entityId;
+    if (!queryEntityId) {
+      const { data } = await createEntity({
+        variables: {
+          entityTypeId: blockProtocolEntityTypes.query.entityTypeId,
+          ownedById: authenticatedUser.accountId as OwnedById,
+          properties: {
+            "https://blockprotocol.org/@hash/types/property-type/query/":
+              integrationData.query,
+          } as QueryProperties,
+        },
+      });
+
+      queryEntityId = data?.createEntity.metadata.recordId.entityId;
+      if (!queryEntityId) {
+        throw new Error("Query entity not created");
+      }
+    }
+
+    const syncPayload: CreateOrUpdateSheetsIntegrationRequest = {
+      audience: integrationData.audience,
+      existingIntegrationEntityId:
+        currentIntegration?.metadata.recordId.entityId,
+      googleAccountId: integrationData.googleAccountId,
+      queryEntityId,
+      spreadsheetId: integrationData.existingFile?.id,
+      newFileName: integrationData.newFileName,
+    };
+
+    await fetch(`${apiOrigin}/integrations/google/sheets`, {
+      method: "POST",
+      credentials: "include",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(syncPayload),
+    });
```

Comment:
Yes, basically. This isn't a great solution, we'd be better off having some proper typed solution to REST endpoints, or just using a GQL resolver. Maybe we can talk about this in the context of flows since creating this integration will probably become creating a Flow that ends up writing entities to a spreadsheet.